Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -60,49 +60,89 @@ extension HTTPClient {
     }
 
     public struct Request {
-        public var version: HTTPVersion
         public var method: HTTPMethod
         public var url: URL
-        public var scheme: String
-        public var host: String
         public var headers: HTTPHeaders
         public var body: Body?
 
-        public init(url: String, version: HTTPVersion = HTTPVersion(major: 1, minor: 1), method: HTTPMethod = .GET, headers: HTTPHeaders = HTTPHeaders(), body: Body? = nil) throws {
+        public init(url: String, method: HTTPMethod = .GET, headers: HTTPHeaders = HTTPHeaders(), body: Body? = nil) throws {
             guard let url = URL(string: url) else {
                 throw HTTPClientError.invalidURL
             }
 
-            try self.init(url: url, version: version, method: method, headers: headers, body: body)
+            self.init(url: url, method: method, headers: headers, body: body)
         }
 
-        public init(url: URL, version: HTTPVersion = HTTPVersion(major: 1, minor: 1), method: HTTPMethod = .GET, headers: HTTPHeaders = HTTPHeaders(), body: Body? = nil) throws {
-            guard let scheme = url.scheme else {
+        public init(url: URL, method: HTTPMethod = .GET, headers: HTTPHeaders = HTTPHeaders(), body: Body? = nil) {
+            self.method = method
+            self.url = url
+            self.headers = headers
+            self.body = body
+        }
+    }
+
+    struct RequestWithHost {
```

Comment:
I'm sorry, I don't quite follow why this is necessary. I must be missing something... Couldn't we treat a `Request`'s `url` property as the source of truth and control access to that?

Reply:
It's not about access control. `URL`'s `host` property is optional, but we require it to be non-optional (so we know where to connect, what default domain to use in cookies and how to redirect using relative `URL`). Before that we would unwrap `host` (or throw) and then store in the request, but this leads to two downsides:
1. We have to keep in sync `host` and `url` properties (and we cannot throw if we change a property, so url property would have to be set using a function, and this is a bit ugly, imho)
2. creating request is a throwing init
To address those two points I propose to introduce new internal type, sort of like "validated request", where user will not be able to change it and `Request` will not be throwing.
Does that make sense?

Reply:
Thanks for the explanation, I see now.

As we control the `URL` so we know its host property is set, can't we force unwrap instead of throwing?

Reply:
What do you mean control the `URL`? `URL` is passed in by the user, so we don't know what will be inside it.

Reply:
Doh. I'm sorry, I see now.  So we swap the throwing initializer for a failed future from `execute()`. I guess that's a better tradeoff although it does mean that one initializer throws and the other doesn't.

Reply:
We can make first initializer optional instead of throwing to match `URL`, wdyt?

Reply:
1. i prefer non-throwing APIs

2. another option here is to define a custom url type in which host is not optional, and with [optional]  initializers from string and URL

Reply:
converted to optional. I'm not sure having our own URL type is the way to go atm, `URL` is used everywhere.

Reply:
> I'm not sure having our own URL type is the way to go atm, URL is used everywhere.

@artemredkin the point was that instead of an internal RequestWithHost you can create and *internal* UrlWithHost (or whatever), as this is really what you want here, not a custom request object. this will help remove all the boilerplate wrapping eg L105-121

Reply:
But how one would pass it further? Handler needs access to URL, host, method and headers. If we don't want to expose `UrlWithHost` we will still need some wrapper to pass both `Request` and `UrlWithHost` to `TaskHandler. Does it make sense?

Reply:
as i was working on demonstrating what i meant above, i believe i ended up with an even simpler and safer approach: https://github.com/swift-server/async-http-client/pull/66

wdyt?